### DISCUSSION TITLE
Send MQTT command to TASMOTA device that is in DeepSleep?

### DISCUSSION THREAD

**Discussion by scotty264** [arendst/Tasmota#20012] [Q&A] [answered]
Folks:
I have a few TASMOTA THB sensors located around the house. One uses a LiIon battery that will run 3 weeks without a recharge. Two others use DC power. All are programmed to use the DeepSleepTimer to wake up and publish MQTT messages every 30 minutes. 

Sometimes, I wish to change some settings but I need to set DeepSleepTime to "0" to do so. But the only way I've found to enter a "DeepSleepTime 0" is to be in the http console during the few seconds it wakes up to report. I usually miss it.

Is there any other way to perform this function?

Perhaps queuing a message with the MQTT broke to send the message when the device reports in?

Thanks in advance.
JQ

**Comment by sfromis** [comment]
Indeed, MQTT is the way; If you publish a retained message on the command topic, this will be picked up by Tasmota on next deep wakeup.
This is already described in the docs: https://tasmota.github.io/docs/DeepSleep/#using-mqtt

**Comment by scotty264** [comment]
Hi: Ah, okay, I was missing the "retained" message part. I thought the MQTT broker would do that. 

But it seems that the -r will keep that message retained forever, so the sensor will keep receiving that command every time it wakes up and publishes a message. If I tell the sensor to enter deepsleep mode again, next time it wakes up, it will receive the retained message again, looping until the message is "un-retained". Is this correct?

If so, is there a means to have the broker send the retained message only one time, like a "one-shot"?

Thank you for the fast response.

Peace and blessings,
JQ

**Reply by sfromis** [reply]
No one-shot, but you can remove the retained message again, by publishing an empty retained message to the same topic. You could even do that from Tasmota itself, command `Publish2`, maybe as part of the retained message itself.

**Reply by scotty264** [reply]
Okay but doesn't that mean I have to keep checking for the device to wake up before publishing an empty retained message?

So, if I comprehend what you are saying, I can also retain a "Publish2" message for the device to execute when it awakes, so that it would turn the retained "deepsleeptimer 0" message off?

Using the retained message feature you mentioned, I wrote a script where I am issuing the retained command then spawning a background process to wait for that sensor to publish its sensor data when it awakes (mosquitto_sub read 1 time) then I issue the empty message to clear the retained message. It works as expected but I'd rather not have to spawn a child process to check.... :(

Thanks again for your quick response.

Peace and blessings,
JQ

**Reply by sfromis** [reply]
Well, the suggestion I made was for how you do **not** need to "keep checking" or having a script doing it. You can simply have the retained message Tasmota gets when waking up include a Publish2 command to remove it again, using `Backlog`

**Reply by scotty264** [reply]
That's what I figured but I'm not an expert on tasmota nor MQTT.

If I comprehend your suggestion, I would issue 2 MQTT commands for the device;

1) `mosquitto_pub -t cmnd/THB-ATTIC/DeepSleepTime -r -m "0"`
       and
2) `mosquitto_pub -t cmnd/THB-ATTIC/Backlog -m "Publish cmnd/THB-ATTIC/DeepSleepTime" -r -n`

Or can this all be combined as:

`mosquitto_pub -t cmnd/THB-ATTIC/Backlog -r -m "DeepSleepTime 0; Publish cmnd/THB-ATTIC/DeepSleepTime"`

Note; I think I want to use `Publish` there and not `Publish2`, correct?

Again, thank you for your time.

Peace and blessings,
JQ

**Reply by sfromis** [reply]
I meant the latter combined one. And you specifically **need** to use `Publish2` to the same topic like `cmnd/THB-ATTIC/Backlog`, as removing a retained message takes publishing an empty message *to the same topic* as it was retained on originally.

**Reply by scotty264** [reply]
Based on your suggestion, I am using:

`mosquitto_pub -h MQTT_BROKER -t "cmnd/THB-ATTIC/Backlog" -r -m "DeepSleepTime 0; Publish2 cmnd/THB-ATTIC/Backlog"`

After executing the above command, the THB-ATTIC sensor comes out of deepsleep as expected. I then reissue a `deepsleeptime 15` and it goes to sleep for ~15 seconds. The next wakeup drops it out of deepsleep mode again (should deepsleep again a few seconds later). So apparently, the retained message for it wasn't cleared by the `Publish2 cmnd/THB-ATTIC/Backlog` command. 

Here is the `DeepSleepTime` query:
`20:31:00.473 CMD: deepsleeptime
20:31:00.484 MQT: stat/THB-ATTIC/RESULT = {"DeepSleepTime":0}`

If I query the sensors `Backlog`, it reports the backlog is empty. 
`20:34:34.395 CMD: backlog
20:34:34.400 MQT: stat/THB-ATTIC/RESULT = {"Backlog":"Empty"}`

I added the THB-ATTIC's restart log at the end of this post. This was after I cleared all topics with `mosquitto_pub -h MQTT_BROKER -t "cmnd/THB-ATTIC" -n -r -d` and a `DeepSleepTime 15` command was issued.

I don't know if I'm battling retained messages, a syntax issue or a bug in the `mosquitto` broker or TASMOTA.

Mosquitto is running under UBUNTU 18.04.
`mosquitto version 1.4.15 (build date Tue, 18 Jun 2019 11:42:22 -0300)`
`mosquitto is an MQTT v3.1.1/v3.1 broker.`

I've stopped `moquitto`, deleted the database then restarted to insure no retained messages exist.

The TASMOTA firmware on all my devices is V:12.4.0, Core: 2.7.4.9. Its the I2C sensor build with AHT and BMP sensor support.

Note that my method of using the script to issue the retained `deepsleeptime 0` message, wait for the sensor to publish and then clear the retained message method works correctly. 

I really would like to use the combined method you suggest.

Thanks.

THB-ATTIC restart log:
`20:36:10.003 MQT: stat/THB-ATTIC/RESULT = {"GPIOs2":{"6656":"TM1637 DIO","6688":"DLP Tx","6720":"DLP Rx","6752":"SSD1351 DC","6784":"XPT2046 CS","6816":"CSE7761 Tx","6848":"CSE7761 Rx","6880":"VL53LXX XSHUT","6912":"MAX7219 CLK","6944":"MAX7219 DIN","6976":"MAX7219 CS","7008":"TFmini+ TX","7040":"TFmini+ RX","7072":"ZC Pulse","7104":"Input","7136":"I2S DOUT","7168":"I2S BCLK","7200":"I2S WS","7232":"I2S DIN","7264":"I2S WS IN","7296":"I2S BCLK IN","7328":"Interrupt","7360":"MCP2515 CS","7392":"HRG15 Tx","7424":"HRG15 Rx","7456":"VINDRIKTNING","7488":"BL0939 Rx","7520":"BL0942 Rx","7552":"HM330X SET","7584":"Heartbeat","7616":"Heartbeat_i","7648":"74x595 SRCLK","7680":"74x595 RCLK","7712":"74x595 OE","7744":"74x595 SER","7776":"SolaxX1 RTS","7808":"Option E","7840":"SDM230 Tx","7872":"SDM230 Rx","7904":"ADC MQ","7936":"CM110x TX","7968":"CM110x RX","8000":"BL6523 Tx","8032":"BL6523 Rx","8064":"ADE7880 IRQ","8096":"Reset","8128":"MS01","8160":"SDIO CMD","8192":"SDIO CLK","8224":"SDIO D0","8256":"SDIO D1","8288":"SDIO D2","8320":"SDIO D3","8352":"Flowrate","8384":"BP5758D Clk","8416":"BP5758D Dat","8448":"SM2335 Clk","8480":"SM2335 Dat","8512":"MP3 Busy","8544":"TM1621 CS","8576":"TM1621 WR","8608":"TM1621 RD","8640":"TM1621 DAT","8672":"Relay_b","8704":"Relay_bi","8736":"I2S MCLK","8768":"ModBr Tx","8800":"ModBr Rx","8832":"ADE7953 RST","8864":"NrgMbs Tx","8896":"NrgMbs Rx","8928":"ADE7953 CS","8960":"Dali RX","8992":"Dali TX","9024":"BP1658CJ Clk","9056":"BP1658CJ Dat","9088":"Dingtian CLK","9120":"Dingtian SDI","9152":"Dingtian Q7","9184":"Dingtian PL","9216":"Dingtian RCK","9248":"LD2410 Tx","9280":"LD2410 Rx","9312":"ModBr Tx Ena","9344":"NrgMbs Tx Ena","9376":"ME007 Tri","9408":"ME007 Rx","9440":"TuyaMCUBr Tx","9472":"TuyaMCUBr Rx","9504":"BioPDU PZEM0XX Tx","9536":"BioPDU PZEM016 Rx","9568":"BioPDU Bit"}}

20:36:11.206 RSL: RESULT = {"SwitchDebounce":50}

20:36:11.479 RSL: RESULT = {"BlinkTime":10}

20:36:12.242 RSL: RESULT = {"MqttLog":2}

20:36:12.763 RSL: STATE = {"Time":"2023-11-14T20:36:12","Uptime":"0T00:00:09","UptimeSec":9,"Heap":21,"SleepMode":"Dynamic","Sleep":250,"LoadAvg":182,"MqttCount":1,"Wifi":{"AP":1,"SSId":"TRUSTNO1","BSSId":"84:D3:43:5D:F9:22","Channel":6,"Mode":"11n","RSSI":82,"Signal":-59,"LinkCount":1,"Downtime":"0T00:00:03"}}

20:36:12.889 MQT: tele/THB-ATTIC/SENSOR = {"Time":"2023-11-14T20:36:12","BMP280":{"Temperature":98.49,"Pressure":738.7},"AHT2X":{"Temperature":95.11,"Humidity":27.41,"DewPoint":56.25},"Global":{"Temperature":95.11,"Humidity":27.41,"DewPoint":4.67,"Pressure":738.7,"SeaPressure":738.8},"PressureUnit":"mmHg","TempUnit":"F"}

20:36:12.918 MQT: stat/THB-ATTIC/RESULT = {"PulseTime2":{"Set":0,"Remaining":0}}

20:36:15.613 MQT: stat/THB-ATTIC/RESULT = {"GPIOs2":{"6656":"TM1637 DIO","6688":"DLP Tx","6720":"DLP Rx","6752":"SSD1351 DC","6784":"XPT2046 CS","6816":"CSE7761 Tx","6848":"CSE7761 Rx","6880":"VL53LXX XSHUT","6912":"MAX7219 CLK","6944":"MAX7219 DIN","6976":"MAX7219 CS","7008":"TFmini+ TX","7040":"TFmini+ RX","7072":"ZC Pulse","7104":"Input","7136":"I2S DOUT","7168":"I2S BCLK","7200":"I2S WS","7232":"I2S DIN","7264":"I2S WS IN","7296":"I2S BCLK IN","7328":"Interrupt","7360":"MCP2515 CS","7392":"HRG15 Tx","7424":"HRG15 Rx","7456":"VINDRIKTNING","7488":"BL0939 Rx","7520":"BL0942 Rx","7552":"HM330X SET","7584":"Heartbeat","7616":"Heartbeat_i","7648":"74x595 SRCLK","7680":"74x595 RCLK","7712":"74x595 OE","7744":"74x595 SER","7776":"SolaxX1 RTS","7808":"Option E","7840":"SDM230 Tx","7872":"SDM230 Rx","7904":"ADC MQ","7936":"CM110x TX","7968":"CM110x RX","8000":"BL6523 Tx","8032":"BL6523 Rx","8064":"ADE7880 IRQ","8096":"Reset","8128":"MS01","8160":"SDIO CMD","8192":"SDIO CLK","8224":"SDIO D0","8256":"SDIO D1","8288":"SDIO D2","8320":"SDIO D3","8352":"Flowrate","8384":"BP5758D Clk","8416":"BP5758D Dat","8448":"SM2335 Clk","8480":"SM2335 Dat","8512":"MP3 Busy","8544":"TM1621 CS","8576":"TM1621 WR","8608":"TM1621 RD","8640":"TM1621 DAT","8672":"Relay_b","8704":"Relay_bi","8736":"I2S MCLK","8768":"ModBr Tx","8800":"ModBr Rx","8832":"ADE7953 RST","8864":"NrgMbs Tx","8896":"NrgMbs Rx","8928":"ADE7953 CS","8960":"Dali RX","8992":"Dali TX","9024":"BP1658CJ Clk","9056":"BP1658CJ Dat","9088":"Dingtian CLK","9120":"Dingtian SDI","9152":"Dingtian Q7","9184":"Dingtian PL","9216":"Dingtian RCK","9248":"LD2410 Tx","9280":"LD2410 Rx","9312":"ModBr Tx Ena","9344":"NrgMbs Tx Ena","9376":"ME007 Tri","9408":"ME007 Rx","9440":"TuyaMCUBr Tx","9472":"TuyaMCUBr Rx","9504":"BioPDU PZEM0XX Tx","9536":"BioPDU PZEM016 Rx","9568":"BioPDU Bit"}}

20:36:16.009 MQT: stat/THB-ATTIC/RESULT = {"SwitchDebounce":50}

20:36:16.296 MQT: stat/THB-ATTIC/RESULT = {"BlinkTime":10}

20:36:16.510 MQT: stat/THB-ATTIC/RESULT = {"MqttLog":2}

20:36:16.759 MQT: stat/THB-ATTIC/RESULT = {"PulseTime2":{"Set":0,"Remaining":0}}

20:36:17.046 MQT: stat/THB-ATTIC/RESULT = {"PulseTime4":{"Set":0,"Remaining":0}}

20:36:17.297 MQT: stat/THB-ATTIC/RESULT = {"PulseTime6":{"Set":0,"Remaining":0}}

20:36:17.547 MQT: stat/THB-ATTIC/RESULT = {"PulseTime8":{"Set":0,"Remaining":0}}

20:36:17.797 MQT: stat/THB-ATTIC/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}

20:36:18.047 MQT: stat/THB-ATTIC/RESULT = {"ButtonDebounce":50}

20:36:18.297 MQT: stat/THB-ATTIC/RESULT = {"BlinkCount":10}

20:36:18.548 MQT: stat/THB-ATTIC/RESULT = {"PulseTime3":{"Set":0,"Remaining":0}}

20:36:18.821 MQT: stat/THB-ATTIC/RESULT = {"PulseTime7":{"Set":0,"Remaining":0}}

20:36:19.051 MQT: stat/THB-ATTIC/RESULT = {"GPIO0":{"640":"I2C SDA"},"GPIO1":{"0":"None"},"GPIO2":{"608":"I2C SCL"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

20:36:19.298 MQT: stat/THB-ATTIC/RESULT = {"PulseTime1":{"Set":0,"Remaining":0}}

20:36:19.549 MQT: stat/THB-ATTIC/RESULT = {"Interlock":"OFF"}

20:36:19.799 MQT: stat/THB-ATTIC/RESULT = {"PulseTime5":{"Set":0,"Remaining":0}}

20:36:22.307 MQT: tele/THB-ATTIC/STATE = {"Time":"2023-11-14T20:36:22","Uptime":"0T00:00:19","UptimeSec":19,"Heap":25,"SleepMode":"Dynamic","Sleep":250,"LoadAvg":5,"MqttCount":2,"Wifi":{"AP":1,"SSId":"TRUSTNO1","BSSId":"84:D3:43:5D:F9:22","Channel":6,"Mode":"11n","RSSI":82,"Signal":-59,"LinkCount":1,"Downtime":"0T00:00:03"}}

20:36:22.334 MQT: tele/THB-ATTIC/SENSOR = {"Time":"2023-11-14T20:36:22","BMP280":{"Temperature":98.46,"Pressure":738.7},"AHT2X":{"Temperature":95.07,"Humidity":27.41,"DewPoint":56.21},"Global":{"Temperature":95.07,"Humidity":27.41,"DewPoint":4.65,"Pressure":738.7,"SeaPressure":738.8},"PressureUnit":"mmHg","TempUnit":"F"}

20:36:49.821 CMD: backlog

20:36:49.827 MQT: stat/THB-ATTIC/RESULT = {"Backlog":"Empty"}

20:36:52.574 CMD: deepsleeptime

20:36:52.581 MQT: stat/THB-ATTIC/RESULT = {"DeepSleepTime":0}

20:37:41.309 MQT: tele/THB-ATTIC/STATE = {"Time":"2023-11-14T20:47:41","Uptime":"0T00:01:09","UptimeSec":69,"Heap":24,"SleepMode":"Dynamic","Sleep":250,"LoadAvg":5,"MqttCount":1,"Wifi":{"AP":1,"SSId":"TRUSTNO1","BSSId":"84:D3:43:5D:F9:22","Channel":6,"Mode":"11n","RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:03"}}

20:37:41.337 MQT: tele/THB-ATTIC/SENSOR = {"Time":"2023-11-14T20:47:41","BMP280":{"Temperature":98.29,"Pressure":738.7},"AHT2X":{"Temperature":94.95,"Humidity":27.23,"DewPoint":55.92},"Global":{"Temperature":94.95,"Humidity":27.23,"DewPoint":4.44,"Pressure":738.7,"SeaPressure":738.9},"PressureUnit":"mmHg","TempUnit":"F"}

20:37:51.315 MQT: tele/THB-ATTIC/STATE = {"Time":"2023-11-14T20:47:51","Uptime":"0T00:01:19","UptimeSec":79,"Heap":24,"SleepMode":"Dynamic","Sleep":250,"LoadAvg":3,"MqttCount":1,"Wifi":{"AP":1,"SSId":"TRUSTNO1","BSSId":"84:D3:43:5D:F9:22","Channel":6,"Mode":"11n","RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:03"}}

20:37:51.342 MQT: tele/THB-ATTIC/SENSOR = {"Time":"2023-11-14T20:47:51","BMP280":{"Temperature":98.35,"Pressure":738.8},"AHT2X":{"Temperature":94.97,"Humidity":27.19,"DewPoint":55.90},"Global":{"Temperature":94.97,"Humidity":27.19,"DewPoint":4.42,"Pressure":738.8,"SeaPressure":738.9},"PressureUnit":"mmHg","TempUnit":"F"}

... continuing tele/THB-ATTIC/SENSOR

... continuing tele/THB-ATTIC/SENSOR

... continuing tele/THB-ATTIC/SENSOR`

**Comment by barbudor** [comment]
> If I query the sensors Backlog, it reports the backlog is empty.
20:36:49.821 CMD: backlog
20:36:49.827 MQT: stat/THB-ATTIC/RESULT = {"Backlog":"Empty"}

This means nothing.
Executing a command in the console is nowhere related to a possible backlog command retained in the broker
To check is there is a retained command use mosquito_sub or any other MQTT tool such as MQTT explorer or MQTTX

**Reply by scotty264** [reply]
Hi:
I've checked using the TASMOTA Device manager and see no retained backlog commands. 

Using MQTT -> Clear obsolete retained LWT's shows none.

Does the command I am using look correct?

Do you have any other suggestion(s) as to why the sensor won't hold the `deepsleeptime`?

Peace and blessings,
JQ

**Reply by sfromis** [reply]
Your log does not really make things clear, as it shows a lot of activity which is **not** from having a restart with the retained command, and then do the publish to reset it. Nothing of that in the log. What it *instead* shows is a lot of commands being executed, but not where they are from, or what you're doing.

Specifically, this *cannot* be from the retained command message, as that has `Backlog`, not lower case `backlog`.
```
20:36:49.821 CMD: backlog
20:36:49.827 MQT: stat/THB-ATTIC/RESULT = {"Backlog":"Empty"}
``` 
Using log level 3 is recommended, to also show command source, and some further details.

Copy/pasting logs here works *much* better if you have triple backtick lines before/after the whole block. Then it gets formatted nicely, without you having to mess with extra blank lines for formatting.

**Reply by scotty264** [reply]
Okay, I'll work with the MQTT loglevel at 3 and get more info.

I didn't think that the commands were case sensitive. I've never seen any indication otherwise.

I've switched to one of the units (THB-CRAWL) that I have in front of me for easier power-cycling. 

Can you confirm that this is the correct command that you had envisioned?

```mosquitto_pub -h MQTT_BROKER -t "cmnd/THB-CRAWL/Backlog" -r -m "DeepSleepTime 0; Publish2 cmnd/THB-CRAWL/Backlog"```

I'll post more soon.

**Reply by sfromis** [reply]
No, the commands are **not** case sensitive. I merely observed that the log you posted does not show the commands being executed, and therefore does not document your case. Like not at all.

And yes, the `mosquitto_pub` command looks right.

**Reply by barbudor** [reply]
> Using MQTT -> Clear obsolete retained LWT's shows none.

As the menu says it only clears the LWT topic

As already said, to check existing retained message use the command `mosquitto_sub -t  +/THB-CRAWL/#` 
That will subscribe to all topics to/from THB-CRAWL
Everything that would appear immediately after running the command is likely to be retained

Unfortunately mosquitto_sub do not tells you which messages are retained or not. That's why I suggested MQTT-Explorer or MQTTX

**Reply by sfromis** [reply]
At least in current iterations, `mosquitto_sub` *can* indicate whether messages are retained or not, by giving an output format option like
`-F "%r %t %p"` appended to the command.
Printed lines have 1/0 for retained or not, the topic, and the payload.

**Reply by scotty264** [reply, marked as answer]
Hi:
Thanks for the responses. And your patience of course. MQTT is a whole other topic that I wasn't planning to get a Ph.d in just to be able to use it effectively. I had these sensors setup and reporting just fine since July but when I wanted to change the telemetry reporting interval, that's where I needed a better solution. 

`sfromis`: The combined MQTT publish command was exactly what I was looking for. Thank you.

`barbudor`: Other than your post, the `mosquitto_sub -t  +/THB-CRAWL/#` isn't anywhere in **this** thread BUT it was exactly what I needed to see the messages that were retained that I didn't know about. That's a great debugging tool! Thank you for that suggestion. 

I was able to see that on the two sensors not adhering to the `deepsleeptime 15` command, that there was a retained `Backlog 2` command that had a `deepsleeptime 0`embedded in it that the sensors were responding to. I was able to clear it and all is working properly.

I now have all three sensors reliably cycling `deepsleeptime 15` telemetry reporting iterations until I break them out with the correct publish command (as sfromis assisted with).

If someone can clarify and confirm for me; I'm concluding that if I publish a retained message to `cmnd/%topic%/[sub-topic]`, even if that [sub-topic] isn't a valid and recognizable command by a TASMOTA device, it will still be retained on the broker and repeatedly fetched by the device until the retain flag is cleared? 

**==================================================**

So to recap for others, the correct "one-shot" MQTT publish command is `mosquitto_pub -h MQTT_BROKER -t "cmnd/THB-CRAWL/Backlog" -r -m "DeepSleepTime 0; Publish2 cmnd/THB-CRAWL/Backlog"`

Use `mosquitto_sub -t  +/THB-CRAWL/#` to debug and find **any** retained messages.

   Where 'THB-CRAWL' is replaced by the sensor's name (topic).

**==================================================**

Thanks again for your support and pateience.

Peace and blessings,
JQ

**Reply by sfromis** [reply]
The MQTT broker does not care about what sub topics are "valid" or not, whatever you publish is simply messages to be forwarded and possibly retained. If you publish an invalid command name, Tasmota will complain about that, but this does not affect what the broker will be doing. Technically, the device is not "fetching" them, retained messages will be forwarded by the broker when a client subscribes to matching topics.

And better be careful with publishing retained messages with commands, and remove them unless you *know* that you want to have them, as they may come back and kick your butt. With Tasmota, you normally do not really need to mess with retained messages, with deep sleep being the main exception, where they are good for breaking into the wakeup cycles.

You do not need `MQTTlog 3` to find what retained messages you have, that option is to get further details of what Tasmota is doing.

**Reply by scotty264** [reply]
Okay, warnings and suggestions heeded.

Thanks again for your assistance.

Peace and blessings,
JQ
